Fix Finder sidebar icon to work as a "template" image

elsiehupp opened the pull request:
Fixes https://github.com/nextcloud/desktop/issues/3695

According to the Apple documentation, "template" images must be exclusively ["monochromatic images that are drawn just using black and transparency"](https://developer.apple.com/library/archive/documentation/General/Conceptual/ExtensibilityPG/Finder.html), but `desktop/theme/colored/Nextcloud-sidebar.svg` looks like this:

<img width="256px" src="https://github.com/nextcloud/desktop/raw/d2b67ffea20b296a512e68f967e77c38a7ff2a4c/theme/colored/Nextcloud-sidebar.svg">

Since this bug popped up without changes in the code or assets in question, my guess is that the current version of Xcode or macOS is being clever and interpreting the white as a color to be inverted, while previously it treated the white as just an opaque transparency mask. This seems to be due to a change in how non-monochrome "template" images are supported with the release of SF Symbols 3. Among other things, SVG images now seem to be natively supported, as best I can tell from the documentation [here](https://developer.apple.com/documentation/uikit/uiimage/creating_custom_symbol_images_for_your_app).

I can't test this theory [because macOS builds are currently broken](https://github.com/nextcloud/desktop/issues/4366), so I'm posting an updated version of `Nextcloud-sidebar.svg` that should work properly as a conventional "template" image without actually put it through a build first.

elsiehupp commented:
GPG signing seems to be broken since I stopped using GitHub's commit-email obfuscation service. I probably need to create a new GPG key.

claucambra commented:
We have a branch (https://github.com/nextcloud/desktop/tree/bugfix/black-sidebar-icon) with a black SVG image and this did not fix the issue unfortunately

claucambra commented:
We do have an update on this bug report #4325 that the issue has been fixed on 12.3, is this the case? I am for now not able to test on this macOS version

elsiehupp commented:
@claucambra do you know how to do a macOS build currently? The existing instructions are broken.

claucambra commented:
> @claucambra do you know how to do a macOS build currently? The existing instructions are broken.

Personally I build it by installing QtCreator and OpenSSL through brew. You'll have to add `-DOPENSSL_ROOT_DIR=/usr/local/opt/openssl` to the initial CMake parameters

claucambra commented:
I'm on macOS 12.3 now and the black sidebar icon bug has now been fixed by Apple (yay!)

Still, we can merge this so that we fully adhere to the guidelines in case things break in the future

elsiehupp commented:
> I'm on macOS 12.3 now and the black sidebar icon bug has now been fixed by Apple (yay!)
> 
> Still, we can merge this so that we fully adhere to the guidelines in case things break in the future

After I got this in my email I updated to 12.3, and it fixed things for me, too.

Looking at [the SF Symbols 3 documentation](https://developer.apple.com/documentation/uikit/uiimage/creating_custom_symbol_images_for_your_app) there are some other aspects of the namespace/specification that might be useful, like idk annotated weights and whatnot. (This would require using the raw SVG rather than rasterizing it.)

Anyway, I made a version of the icon that tweaks the geometry slightly in order to be more in line with SF Symbols. The first two are the tweaked version of the icon, and the second two are a slightly rationalized version of the existing icon.

<img alt="nextcloud-sidebar-sf-symbols-black" width="32" src="https://user-images.githubusercontent.com/9206310/160139971-b099e733-eecd-4f9c-aeef-bac975ed207f.svg"><img alt="nextcloud-sidebar-sf-symbols-white" width="32" src="https://user-images.githubusercontent.com/9206310/160137261-15dc8fc6-e366-4a46-aea1-6d2c9c410da6.svg"><img alt="nextcloud-sidebar-original-black" width="32" src="https://user-images.githubusercontent.com/9206310/160140028-6acca45a-44ef-4659-871b-5c4642658f42.svg"><img alt="nextcloud-sidebar-original-white" width="32" src="https://user-images.githubusercontent.com/9206310/160140048-7f47bea3-8df3-4de8-ac94-da2a2cd3d6ad.svg">

And here's a screenshot of my Finder sidebar to show how the existing icon looks next to the standard SF Symbols icons:

<img width="146" alt="finder-sidebar" src="https://user-images.githubusercontent.com/9206310/160140359-ab41e675-a4d9-46ac-94e1-7f2d5bcab82f.png">

The main thing I changed is making the circles tangent to each other at the centerlines of their strokes, which subsequently allowed me to change the line weight. An SF Symbols regular-size/medium-weight icon seems to have roughly 7px line weight within a 128px canvas. I also tweaked the geometry of the circles to make them have simple integer ratios: the large circle has a diameter of 49px, and the smaller circles have a diameter of 28px, both numbers which are intentionally integer multiples of 7px (the line weight).

FWIW if I were to get carried away I could thoroughly "Mac-ify" the rest of the monochrome icons for the macOS build, but that really has nothing to do with this pull request (and neither, really, does the "rationalized" SF-Symbols version of the sidebar icon).

Back on the original topic, while I did use stroke width for the versions of the icon I attached to this comment, I could with not too much difficulty make visually indistinguishable versions that use masking instead, so that the SVG could completely eliminate style information.

As it is, the rationalized version of the icon I have attached dramatically simplifies the SVG markup, from this:

```xml
<?xml version="1.0" encoding="utf-8"?>
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024">
<!-- This SVG must be black in order to render correctly as a macOS "template" image. -->
<path d="M513.6,266.9c-106.7,0-196.3,73.1-223.6,171.3
	c-23.9-52.6-76.5-89.8-137.6-89.8C69.2,348.5,0.8,416.9,0.8,500c0,83.2,68.4,151.6,151.6,151.6c61.1,0,113.7-37.3,137.5-89.9
	c27.4,98.3,117,171.4,223.7,171.4c106.1,0,195.2-72.3,223.2-169.7c24.3,51.6,76.1,88.2,136.5,88.2c83.2,0,151.6-68.4,151.6-151.6
	c0-83.2-68.4-151.6-151.6-151.6c-60.4,0-112.2,36.6-136.5,88.2C708.8,339.2,619.7,266.9,513.6,266.9L513.6,266.9z M513.6,355.9
	c80.1,0,144.1,64,144.1,144.1c0,80.1-64,144.2-144.1,144.1c-80.1,0-144.1-64-144.1-144.1C369.5,419.9,433.5,355.9,513.6,355.9
	L513.6,355.9z M152.3,437.4c35.1,0,62.6,27.5,62.6,62.6c0,35.1-27.5,62.6-62.6,62.6c-35.1,0-62.6-27.5-62.6-62.6
	C89.8,464.9,117.2,437.4,152.3,437.4L152.3,437.4z M873.2,437.4c35.1,0,62.6,27.5,62.6,62.6c0,35.1-27.5,62.6-62.6,62.6
	c-35.1,0-62.6-27.5-62.6-62.6C810.6,464.9,838.1,437.4,873.2,437.4L873.2,437.4z"/>
</svg>
```

(Which already has cruft removed from the original...)

To this:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" style="fill:none;stroke:black;stroke-width:11px;">
    <circle id="right" cx="109" cy="64" r="13.5"/>
    <circle id="center" cx="64" cy="64" r="23.5"/>
    <circle id="left" cx="19" cy="64" r="13.5"/>
</svg>
```

With masking I would still be able to make the icon entirely out of circles, rather than using béziers, like the current version does.

elsiehupp commented:
Here's the original icon using masks:

<img alt="nextcloud-sidebar-original-mask-black" width="32" src="https://user-images.githubusercontent.com/9206310/160145693-6742dbf3-8366-4983-9200-9c3dd930af71.svg"><img alt="nextcloud-sidebar-original-mask-white" width="32" src="https://user-images.githubusercontent.com/9206310/160145708-ba2add20-f30b-408c-be74-061ac29eb85c.svg">

```svg
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <circle id="left" cx="19" cy="64" r="19" mask="url(#left-clip)"/>
    <circle id="center" cx="64" cy="64" r="29" mask="url(#center-clip)"/>
    <circle id="right" cx="109" cy="64" r="19" mask="url(#right-clip)"/>
    <defs>
        <mask id="left-clip">
            <rect x="0" y="0" width="100%" height="100%" fill="white"/>
            <circle cx="19" cy="64" r="8"/>
        </mask>
        <mask id="center-clip">
            <rect x="0" y="0" width="100%" height="100%" fill="white"/>
            <circle cx="64" cy="64" r="18"/>
        </mask>
        <mask id="right-clip">
            <rect x="0" y="0" width="100%" height="100%" fill="white"/>
            <circle cx="109" cy="64" r="8"/>
        </mask>
    </defs>
</svg>
```

elsiehupp commented:
I wouldn't exactly say it's less complicated than the bézier version, but it's definitely more human-readable!

elsiehupp commented:
This is a slightly different version of the masked SVG, where there is only one mask for the whole image, rather than one mask per circle:

<img alt="nextcloud-sidebar-original-mask-combo-black" width="32" src="https://user-images.githubusercontent.com/9206310/160147657-a655f1f2-418b-4354-885f-699dce192f75.svg"><img alt="nextcloud-sidebar-original-mask-combo-white" width="32" src="https://user-images.githubusercontent.com/9206310/160147679-eefcc831-b59a-49ba-b4cc-d3c9730955b9.svg">

```svg
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <g mask="url(#composite-mask)">
        <circle id="left-circle" cx="19" cy="64" r="19"/>
        <circle id="center-circle" cx="64" cy="64" r="29"/>
        <circle id="right-circle" cx="109" cy="64" r="19"/>
    </g>
    <defs>
        <mask id="composite-mask">
            <rect id="mask-canvas" x="0" y="0" width="100%" height="100%" fill="white"/>
            <circle id="left-circle-cutout" cx="19" cy="64" r="8"/>
            <circle id="center-circle-cutout" cx="64" cy="64" r="18"/>
            <circle id="right-circle-cutout" cx="109" cy="64" r="8"/>
        </mask>
    </defs>
</svg>
```

I also gave elements IDs even where unnecessary, just to annotate what's what.

I think this version is a bit easier to grok than the previous one. What do you think? I need to replace my commit with one that has the correct GPG signature anyway, so I can easily use any of these for the pull request.

claucambra commented:
> > I'm on macOS 12.3 now and the black sidebar icon bug has now been fixed by Apple (yay!)
> > Still, we can merge this so that we fully adhere to the guidelines in case things break in the future
> 
> After I got this in my email I updated to 12.3, and it fixed things for me, too.
> 
> Looking at [the SF Symbols 3 documentation](https://developer.apple.com/documentation/uikit/uiimage/creating_custom_symbol_images_for_your_app) there are some other aspects of the namespace/specification that might be useful, like idk annotated weights and whatnot. (This would require using the raw SVG rather than rasterizing it.)
> 
> Anyway, I made a version of the icon that tweaks the geometry slightly in order to be more in line with SF Symbols. The first two are the tweaked version of the icon, and the second two are a slightly rationalized version of the existing icon.
> 
> <img alt="nextcloud-sidebar-sf-symbols-black" width="32" src="https://user-images.githubusercontent.com/9206310/160139971-b099e733-eecd-4f9c-aeef-bac975ed207f.svg"><img alt="nextcloud-sidebar-sf-symbols-white" width="32" src="https://user-images.githubusercontent.com/9206310/160137261-15dc8fc6-e366-4a46-aea1-6d2c9c410da6.svg"><img alt="nextcloud-sidebar-original-black" width="32" src="https://user-images.githubusercontent.com/9206310/160140028-6acca45a-44ef-4659-871b-5c4642658f42.svg"><img alt="nextcloud-sidebar-original-white" width="32" src="https://user-images.githubusercontent.com/9206310/160140048-7f47bea3-8df3-4de8-ac94-da2a2cd3d6ad.svg">
> 
> And here's a screenshot of my Finder sidebar to show how the existing icon looks next to the standard SF Symbols icons:
> 
> <img alt="finder-sidebar" width="146" src="https://user-images.githubusercontent.com/9206310/160140359-ab41e675-a4d9-46ac-94e1-7f2d5bcab82f.png">
> 
> The main thing I changed is making the circles tangent to each other at the centerlines of their strokes, which subsequently allowed me to change the line weight. An SF Symbols regular-size/medium-weight icon seems to have roughly 7px line weight within a 128px canvas. I also tweaked the geometry of the circles to make them have simple integer ratios: the large circle has a diameter of 49px, and the smaller circles have a diameter of 28px, both numbers which are intentionally integer multiples of 7px (the line weight).
> 
> FWIW if I were to get carried away I could thoroughly "Mac-ify" the rest of the monochrome icons for the macOS build, but that really has nothing to do with this pull request (and neither, really, does the "rationalized" SF-Symbols version of the sidebar icon).
> 
> Back on the original topic, while I did use stroke width for the versions of the icon I attached to this comment, I could with not too much difficulty make visually indistinguishable versions that use masking instead, so that the SVG could completely eliminate style information.
> 
> As it is, the rationalized version of the icon I have attached dramatically simplifies the SVG markup, from this:
> 
> ```
> <?xml version="1.0" encoding="utf-8"?>
> <svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024">
> <!-- This SVG must be black in order to render correctly as a macOS "template" image. -->
> <path d="M513.6,266.9c-106.7,0-196.3,73.1-223.6,171.3
> 	c-23.9-52.6-76.5-89.8-137.6-89.8C69.2,348.5,0.8,416.9,0.8,500c0,83.2,68.4,151.6,151.6,151.6c61.1,0,113.7-37.3,137.5-89.9
> 	c27.4,98.3,117,171.4,223.7,171.4c106.1,0,195.2-72.3,223.2-169.7c24.3,51.6,76.1,88.2,136.5,88.2c83.2,0,151.6-68.4,151.6-151.6
> 	c0-83.2-68.4-151.6-151.6-151.6c-60.4,0-112.2,36.6-136.5,88.2C708.8,339.2,619.7,266.9,513.6,266.9L513.6,266.9z M513.6,355.9
> 	c80.1,0,144.1,64,144.1,144.1c0,80.1-64,144.2-144.1,144.1c-80.1,0-144.1-64-144.1-144.1C369.5,419.9,433.5,355.9,513.6,355.9
> 	L513.6,355.9z M152.3,437.4c35.1,0,62.6,27.5,62.6,62.6c0,35.1-27.5,62.6-62.6,62.6c-35.1,0-62.6-27.5-62.6-62.6
> 	C89.8,464.9,117.2,437.4,152.3,437.4L152.3,437.4z M873.2,437.4c35.1,0,62.6,27.5,62.6,62.6c0,35.1-27.5,62.6-62.6,62.6
> 	c-35.1,0-62.6-27.5-62.6-62.6C810.6,464.9,838.1,437.4,873.2,437.4L873.2,437.4z"/>
> </svg>
> ```
> 
> (Which already has cruft removed from the original...)
> 
> To this:
> 
> ```
> <?xml version="1.0" encoding="UTF-8" standalone="no"?>
> <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
> <svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" style="fill:none;stroke:black;stroke-width:11px;">
>     <circle id="right" cx="109" cy="64" r="13.5"/>
>     <circle id="center" cx="64" cy="64" r="23.5"/>
>     <circle id="left" cx="19" cy="64" r="13.5"/>
> </svg>
> ```
> 
> With masking I would still be able to make the icon entirely out of circles, rather than using béziers, like the current version does.

I like the first (skinnier) icon a lot, it definitely integrates with the other sidebar icons better than our current one does. @jancborchardt thoughts? There's nothing major to change here code-wise so this feels like more of a design discussion :)

elsiehupp commented:
And here's a version that I think is even more legible even if it's a bit verbose:

```svg
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <rect id="left-circle-canvas" x="0" y="0" width="100%" height="100%" mask="url(#left-circle-mask)"/>
    <rect id="center-circle-canvas" x="0" y="0" width="100%" height="100%" mask="url(#center-circle-mask)"/>
    <rect id="right-circle-canvas" x="0" y="0" width="100%" height="100%" mask="url(#right-circle-mask)"/>
    <defs>
        <mask id="left-circle-mask">
            <circle id="left-circle-outer" cx="19" cy="64" r="19" fill="white"/>
            <circle id="left-circle-inner" cx="19" cy="64" r="8" fill="black"/>
        <mask id="center-circle-mask">
        </mask>
            <circle id="center-circle-outer" cx="64" cy="64" r="29" fill="white"/>
            <circle id="center-circle-inner" cx="64" cy="64" r="18" fill="black"/>
        <mask id="right-circle-mask">
        </mask>
            <circle id="right-circle-outer" cx="109" cy="64" r="19" fill="white"/>
            <circle id="right-circle-inner" cx="109" cy="64" r="8" fill="black"/>
        </mask>
    </defs>
</svg>
```

Unless there's a strong reason to use fill rather than strokes, though, I think just using the circles with stroke widths would be simplest:

```svg
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" fill="none" stroke="black" stroke-width="7px">
    <circle id="right" cx="109" cy="64" r="13.5"/>
    <circle id="center" cx="64" cy="64" r="23.5"/>
    <circle id="left" cx="19" cy="64" r="13.5"/>
</svg>
```

(IIRC the main reason to use fill rather than strokes is that GTK "symbolic" icons don't support strokes... but Nextcloud uses Qt rather than GTK, and anyway this particular asset is going directly through Xcode, so it probably doesn't matter.)

elsiehupp commented:
And here's the equivalent version of the SF-Symbols-style icon:

```svg
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 128 128" version="1.1" xmlns="http://www.w3.org/2000/svg" fill="none" stroke="black" stroke-width="7px">
    <circle id="right" cx="102.5" cy="64" r="14"/>
    <circle id="center" cx="64" cy="64" r="24.5"/>
    <circle id="left" cx="25.5" cy="64" r="14"/>
</svg>
```

mgallien commented:
@elsiehupp thanks for your contribution
once we reach a state where this PR can be merged we will wait a bit more due to the preparation of the next feature release

elsiehupp commented:
/rebase

jancborchardt reviewed (approved):
Looks good! :)

elsiehupp commented:
Since this PR is still pending, I changed it to use the SF Symbols lineweights. I also changed the filename to emphasize that the icon is only used on macOS, and I imagine that if you were to merge the corresponding change to `CMakeLists.txt` you would need to change the corresponding filenames in all the OEM builds, as well.

elsiehupp commented:
And I fixed my GPG key! Yay!
